src/tokenizedSubaccounts/LevBasisTSA.sol

```diff
@@ -0,0 +1,476 @@
+      _verifyWithdrawAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.tradeModule)) {
+      _verifyTradeAction(action, tsaAddresses);
+    } else {
+      revert LBT_InvalidModule();
+    }
+  }
+
+  /////////////
+  // Trading //
+  /////////////
+
+  function _verifyTradeAction(IMatching.Action memory action, BaseTSAAddresses memory tsaAddresses) internal {
+    LBTSAStorage memory $ = _getLBTSAStorage();
+
+    ITradeModule.TradeData memory tradeData = abi.decode(action.data, (ITradeModule.TradeData));
+
+    // either base or perp
+    TradeHelperVars memory tradeHelperVars =
+      _getTradeHelperVars(tradeData.asset, address(tsaAddresses.wrappedDepositAsset), address($.perpAsset));
+
+    // Fees
+    if (tradeHelperVars.isBaseTrade) {
+      _verifyCollateralTradeFee(tradeData.worstFee, tradeHelperVars.basePrice);
+    } else {
+      _verifyPerpTradeFee(tradeData.worstFee, tradeHelperVars.perpPrice);
+    }
+
+    int amtDelta = tradeData.isBid ? tradeData.desiredAmount : -tradeData.desiredAmount;
+
+    _verifyTradeDelta(tradeHelperVars, amtDelta);
+    _verifyTradeLeverage(tradeHelperVars, amtDelta);
+    _verifyEmaMarkLoss(tradeData, tradeHelperVars);
+  }
+
+  function _verifyWithdrawAction(IMatching.Action memory action, BaseTSAAddresses memory tsaAddresses) internal view {
+    LBTSAStorage storage $ = _getLBTSAStorage();
+    IWithdrawalModule.WithdrawalData memory withdrawalData = abi.decode(action.data, (IWithdrawalModule.WithdrawalData));
+
+    require(withdrawalData.asset == address(tsaAddresses.wrappedDepositAsset), LBT_InvalidWithdrawAsset());
+
+    TradeHelperVars memory tradeHelperVars =
+      _getTradeHelperVars(withdrawalData.asset, address(tsaAddresses.wrappedDepositAsset), address($.perpAsset));
+
+    _verifyTradeDelta(tradeHelperVars, -int(withdrawalData.assetAmount));
+    _verifyTradeLeverage(tradeHelperVars, -int(withdrawalData.assetAmount));
+  }
+
+  //////////////////////////
+  // Verification Helpers //
+  //////////////////////////
+
+  function getTradeHelperVars(address asset) external view returns (TradeHelperVars memory tradeHelperVars) {
+    tradeHelperVars = _getTradeHelperVars(
+      asset, address(getBaseTSAAddresses().wrappedDepositAsset), address(_getLBTSAStorage().perpAsset)
+    );
+  }
+
+  function _getTradeHelperVars(address asset, address wrappedDepositAsset, address perpAsset)
+    internal
+    view
+    returns (TradeHelperVars memory tradeHelperVars)
+  {
+    tradeHelperVars.isBaseTrade = asset == wrappedDepositAsset;
+
+    require(tradeHelperVars.isBaseTrade || asset == perpAsset, LBT_InvalidTradeAsset());
+
+    tradeHelperVars.basePrice = _getBasePrice();
+    tradeHelperVars.perpPrice = _getPerpPrice();
+    (tradeHelperVars.perpPosition, tradeHelperVars.baseBalance, tradeHelperVars.cashBalance) = _getSubAccountStats();
+    tradeHelperVars.underlyingBase = _getConvertedMtM();
```

Comment:
I noticed this function actually reverts if mTm < 0. I think it was initially done to prevent deposits / withdrawals, but now that it's used in this generic helper, should we move this outside. 

E.g. would this block any valid action / deleveraging when vault is in a dangerous situation?

```
if (convertedMtM < 0 || margin < 0) {
  revert CMTSA_PositionInsolvent();
}
```

Reply:
If MtM < 0, its actually in the vault's interest to get liquidated before depositing, since you'd be paying off insolvency using new funds. 

Will remove the margin check though, since if it is flagged, the action will be blocked anyways.